### ISSUE TITLE
Invalid param "not an object" when using {{#tests}} on a non-object

### ISSUE THREAD

**Issue by Niols** [rgrinberg/ocaml-mustache#37]
Hello,

I was trying to use Mustache to render a string only when it is non-null. This should essentially be the following template:

```
Hello
{{#key}}- The key exists: {{key}}{{/key}}
Good bye
```

It works as expected with the Ruby implementation of Mustache in both cases. It does not however work with the OCaml implementation in the case where `key` exists. I get `Mustache_types.Invalid_param("str. not an object")`.

My guess is that Mustache is happy with the test `{{#key}}` and puts the JSON for `key` in the context. When later rendering `{{key}}` it fails because the context is not an object. It's actually a bit more than a guess:

In `render_fmt`, when rendering sections, we have:

https://github.com/rgrinberg/ocaml-mustache/blob/0333e46e1d010d1dabc1397a6d2841145fa62076/lib/mustache.ml#L324

where `add_context` is:

https://github.com/rgrinberg/ocaml-mustache/blob/0333e46e1d010d1dabc1397a6d2841145fa62076/lib/mustache.ml#L268-L272

which means that when `ctx` is not an object, the context becomes this non-object. And when we try to render something afterwards, we have a failure. I would rather have something like:

```ocaml
    let add_context ctx js =
      match (ctx, js) with
      | `O ctx_o, `O js_o -> `O (ctx_o @ js_o)
      | _, `O _ -> js
      | _, _ -> ctx
    in
```

Does that seem reasonable to you?

**Comment by rgrinberg** [comment]
Yeah, that seems reasonable to me.

**Comment by tsileo** [comment]
Hey, is there any way to have the fix in a new opam release?

Should I make a PR  (with the fix from @Niols fork)?

Thanks!

**Comment by gasche** [comment]
I hit this issue independently and proposed a fix in #49. My proposal is more invasive, but it more clearly corresponds to the specification, and there is no regression in existing tests.

**Comment by gasche** [comment]
(The fix in #49 has been merged, so I think this issue and the other PR #41 could be closed.)
